src/acl/Random.cc

```diff
@@ -92,11 +93,7 @@ ACLRandom::parse()
 int
 ACLRandom::match(ACLChecklist *)
 {
-    // make up the random value.
-    // The fixed-value default seed is fine because we are
-    // actually matching whether the random value is above
-    // or below the configured threshold ratio.
-    static std::mt19937 mt;
+    static std::mt19937 mt(Seed32());
```

Comment:
I think that the justification for the default seed is a little flawed - basically it says that we do not care for seed by the 'random' ACL definition. Strictly speaking, we are matching whether the random value is below or equal the configured ratio (not above). Moreover, I don't think that random ACL should produce deterministic results in different Squid instances - that is why we do need a non-default seed here.

Reply:
It is not clear what invariant that justification is trying to uphold. "Fine" for what purpose? To paraphrase, the comment is saying that "a one-for-all seed does not violate X because we are..." but it is not clear what that X actually is!

AFAICT, in your analysis, you define X as "random ACL should produce non-deterministic results in different Squid instances" but that (alleged) definition does not quite make sense to me because I do not understand what "deterministic in different instances" really means. Please note that all pseudo-random generator algorithms are [deterministic](https://en.wikipedia.org/wiki/Deterministic_system#In_computer_science).

Furthermore, one can predict the next value of any pseudo-random number generator after observing the previous N values (N depends on the generation algorithm; it is about 624 for std::mt19937 AFAICT). Thus, we cannot say that X is simply "the outcome of the next random ACL test by a Squid instance cannot be easily predicted" because we cannot uphold that standard in many realistic Squid configurations where an attacker can observe the effects of the ACL evaluation (e.g., whether a particular cache_peer has been chosen).

I think that X should be (at least) the conjunction of the following invariants (listed here with their applicability-to-Squid rationale):

1. The outcome of n-th random test of ACL a1 should not correlate with that of ACL a2. The outcome sequences of two different ACLs should be independent. Rationale: If I have two random ACLs, I do not want them to look at transactions in a lock-step fashion. Such synchronization often leads to load imbalances and other problems (e.g., all requests going to parent P1 end up cached on disk D1 resulting in P1 overload if cache D1 is emptied).

2. The outcome of n-th random ACL test of acl A in Squid worker W1 should not correlate with that in worker W2. The outcome sequences of same ACLs in two different Squid workers should be independent. Rationale: I do not want two SMP workers to look at transactions in a lock-step fashion. Again, such synchronization often leads to load imbalances and other problems (e.g., all workers selecting the same parent cache at the same time). Same for two "identical" Squid instances in a farm instead of SMP workers.

3. The outcome of first few random ACL test(s) cannot be easily predicted by an outside observer. Rationale: If I find a way to crash Squid, I should not be able to then reliably predict how that Squid will route my first-after-the-crash request(s), even if I know that Squid configuration outline.

Technically, invariant 3 can be viewed as a special case of invariant 2 by considering sequentially started Squids as two different Squid instances. I listed both invariants to provide two different/orthogonal use cases.

The current official implementation upholds invariant 1 but violates invariants 2 and 3.

:warning: I do not recommend fixing these violations in this PR.

Reply:
> do not understand what "deterministic in different instances" really means

Actually I meant that outcomes of a random ACL will be the same indifferent instances, i.e., having determined them in one instance we can predict them in another instance (assuming, for simplicity, that a single random ACL is configured).

>  I do not recommend fixing these violations in this PR

I do not mind.